### PR TITLE
Kimmy/code coverage

### PR DESCRIPTION
**Pull request by jya-kmu**
Enable ctest with code coverage with github action.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -141,7 +141,8 @@ endif()
 # Build options
 #------------------------------------------------------------------------------
 option(BUILD_SHARED_LIBS "Build with shared libraries." ON)
-option(HERMES_INTERCEPT_IO "Use GOTCHA as IO interception tool" ON)
+option(BUILD_TESTING "Build testing." ON)
```
Comment (ChristopherHogan) [inline CMakeLists.txt:144]: According to [the documentation](https://cmake.org/cmake/help/v3.10/module/CTest.html?highlight=build_testing), the `CTest` module "automatically creates a `BUILD_TESTING` option that selects whether to enable testing support (`ON` by default)."

File: CMakeLists.txt [hunk 2]
```diff
@@ -316,11 +337,9 @@ endif()
 #-----------------------------------------------------------------------------
 # Testing
 #-----------------------------------------------------------------------------
-if(CMAKE_PROJECT_NAME STREQUAL HERMES)
+if(NOT HERMES_EXTERNALLY_CONFIGURED AND BUILD_TESTING)
+  enable_testing()
```
Comment (ChristopherHogan) [inline CMakeLists.txt:341]: From [the documentation](https://cmake.org/cmake/help/v3.10/command/add_test.html#command:add_test):
> CMake will generate tests only if the enable_testing() command has been invoked. **The CTest module invokes the command automatically when the BUILD_TESTING option is ON.**

File: CMakeLists.txt [hunk 3]
```diff
@@ -316,11 +337,9 @@ endif()
 #-----------------------------------------------------------------------------
 # Testing
 #-----------------------------------------------------------------------------
-if(CMAKE_PROJECT_NAME STREQUAL HERMES)
+if(NOT HERMES_EXTERNALLY_CONFIGURED AND BUILD_TESTING)
```
Comment (ChristopherHogan) [inline CMakeLists.txt:340]: Can you explain what the `HERMES_EXTERNALLY_CONFIGURED` option is for? I've never found it useful for anything.

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -1,4 +1,5 @@
 # This is a basic workflow to help you get started with Actions
+# Dashborad location https://cdash.hdfgroup.org/index.php?project=HERMES
```
Comment (ChristopherHogan) [inline .github/workflows/main.yml:2]: Typo, should be "Dashboard"

File: CMakeLists.txt [hunk 5]
```diff
@@ -153,6 +154,10 @@ option(HERMES_DEBUG_HEAP "Store Heap debug information for visualization." OFF)
 option(HERMES_BUILD_BENCHMARKS "Build the Hermes benchmark suite." OFF)
 option(HERMES_ENABLE_TIMING "Turn on timing of selected functions." OFF)
 
+if(NOT HERMES_EXTERNALLY_CONFIGURED)
+  option(HERMES_ENABLE_COVERAGE "Enable coverage." OFF)
```
Comment (ChristopherHogan) [inline CMakeLists.txt:158]: Why is this in an `if` block.

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -25,11 +26,15 @@ foreach(program ${API_TESTS})
     "${CMAKE_BINARY_DIR}/bin/${program}")
   set_tests_properties("Test${program}" PROPERTIES ENVIRONMENT
     LSAN_OPTIONS=suppressions=${CMAKE_CURRENT_SOURCE_DIR}/data/asan.supp)
+  if(HERMES_ENABLE_COVERAGE)
+    set_coverage_flags(${program})
```
Comment (ChristopherHogan) [inline test/CMakeLists.txt:30]: I don't think we need the coverage flags on each test. We're only interested in coverage in the library, not which lines get executed in the tests themselves. Right?

File: test/script/ref_script.cmake [hunk 7]
```diff
@@ -0,0 +1,104 @@
+# Usage:
```
Comment (ChristopherHogan) [inline test/script/ref_script.cmake:1]: What is this file for? Is it left over from testing/debugging?

**Comment by ChristopherHogan** [comment]
When I examine the most recent coverage report generated [here](https://cdash.hdfgroup.org/viewCoverage.php?buildid=58828&status=6&dir=src) it only lists 4 files. Where are the rest?

File: CMakeLists.txt (hunk 1 above)
Comment (jya-kmu) [reply CMakeLists.txt:144]: Yes, I was thinking to rename it to `HERMES_BUILD_TESTING`, then there will be no confusion.

File: test/script/ref_script.cmake (hunk 7 above)
Comment (jya-kmu) [reply test/script/ref_script.cmake:1]: That's for ppl who wants to do ctest locally from a script, which can be used as reference for connivence.

File: test/CMakeLists.txt (hunk 6 above)
Comment (jya-kmu) [reply test/CMakeLists.txt:30]: I have seen both cases, including test or not. Either is fine.

**Comment by jya-kmu** [comment]
> When I examine the most recent coverage report generated [here](https://cdash.hdfgroup.org/viewCoverage.php?buildid=58828&status=6&dir=src) it only lists 4 files. Where are the rest?

Check this one (https://cdash.hdfgroup.org/viewCoverage.php?buildid=58916&status=6&dir=src). The reason is test name inconsistency.

File: CMakeLists.txt (hunk 3 above)
Comment (jya-kmu) [reply CMakeLists.txt:340]: Coverage is done only when we are building hermes, but not when using it as a dependency of another library.

**Pull request closed without merging**